compile-time-programming/requires-expressions.md

```diff
@@ -0,0 +1,102 @@
+  </tr>
+</table>
+
+## Motivation
+
+Requires-expressions allow a developer to perform compile-time evaluation on the validity of other expressions. These are fundamental to the ability to write concepts.[[Compile-time programming: concepts]][1]
+
+## Topic introduction
+
+Requires-expressions are compile-time predicates which evaluate to true when their specified set of expressions are all valid for a given set of inputs.
+
+## Foundational knowledge: Writing requires-expressions
+
+### Background/Required Knowledge
+
+A student is able to:
+
+* Write and use a function template [[Compile-time programming: function templates]][2]
+* Differentiate between expressions and statements
+
+It is helpful if:
+
+* A student is aware of the constraints imposed by their template definitions
+
+### Student outcomes
+
+A student should be able to:
+
+1. Write a simple-requirement to assert the validity of an expression
+2. Write a type-requirement to check the existence of a type by its identifier
+3. Write a compound-requirement to test the resulting type of an expression
+4. Write a nested-requirement to test the constexpr value of an operation, as opposed to just the semantic validity
+5. Use a requires-expression within a concept, requires-clause, or `if constexpr` condition
+
+### Caveats
+
+To require expressions which evaluate to a boolean value, such as `sizeof(t) == 4`, to only be satisfied when the result of the expression is `true`, you must use a nested-requirement, such as `requires sizeof(t) == 4;`. Omitting the `requires` results in a simple-requirement, which is satisfied based purely on semantic validity, not on the result of the operation.
+
+### Points to cover
+
+* All requires-expression requirements terminate with a semi-colon.
+* simple-requirements are used to check the semantics of an expression.
+* nested-requirements are introduced with `requires` and primarily used to check the result of an expression computable by the compiler, including concepts or other requires-expressions.
+* type-requirements are introduced with `typename` and used to verify the existence of a type with a particular identifier.
+* compound-requirements are enclosed in braces and can be used to check the resulting type of an expression.
+* Parameters should be specified without cv-qualifiers, as though passed by-value.
+* Checks are performed by the compiler, not at run time.
+
+## Advanced: Advanced requirements
+
+### Background/required knowledge
+
+* All of the above.
+* Knowledge of `noexcept`
+
+A student is able to:
+
+* Write a concept [[Compile-time programming: concepts]][1]
+
+### Student outcomes
+
+A student should be able to:
+
+1. Write compound-requirements which test the `noexcept`ness of an expression.
+2. Use a concept as the target of a compound-requirement.
+
+### Caveats
+
+### Points to cover
+
+* Compound-requirements allow the optional ability to test whether an expression is marked as `noexcept`, by using the `noexcept` keyword.
```

Comment:
Yikes! I would be much more specific here. This is another of my "Concepts As She Is Spoke" examples. https://godbolt.org/z/p3bWM9

    template<class T>
    concept True = requires(T t) {
        noexcept( t.foo() );
    };
    struct S { int foo() noexcept(false); };
    static_assert(True<S>);

You'll notice a pattern: simple-requirements are the devil and should be avoided at all costs. (But I'm not necessarily saying that that guideline belongs in _this_ document.)

Reply:
I am not clear what suggestion, if any, is made here. I'd hesitate to fill these documents with all sorts of edge cases, but this may warrant inclusion?

Reply:
I think just `s/using the noexcept keyword/using a trailing noexcept keyword/` would be sufficient. Arguably the example below is sufficient; but the example is long enough that the reader's eyes may glaze over before getting to the sneaky trailing-noexcept part.